flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -741,6 +763,18 @@ private void establishJobManagerConnection(JobID jobId, JobMasterGateway jobMast
 			jobManagerTable.put(jobId, associateWithJobManager(jobMasterGateway, jobManagerLeaderId, registrationSuccess.getBlobPort()));
 		}
 
+		heartbeatManager.monitorTarget(registrationSuccess.getResourceID(), new HeartbeatTarget() {
+			@Override
+			public void sendHeartbeat(ResourceID resourceID, Object payload) {
+				jobMasterGateway.heartbeatFromTaskManager(resourceID, payload);
+			}
+
+			@Override
+			public void requestHeartbeat(ResourceID resourceID, Object payload) {
+				throw new UnsupportedOperationException("Should never call requestHeartbeat in task manager.");
```

Comment:
If I understand correctly the TM can request the heartbeat to itself and the JM also can send the heartbeat to itself? 
If it is true, the **requestHeartbeat** will never be called from current logic in TM side. Or I can just remove the **UnsupportedOperationException** to not limitation and leave this method empty currently?